test/src/org/labkey/test/tests/targetedms/InstrumentSchedulingTest.java

```diff
@@ -0,0 +1,221 @@
+        ));
+        List<Map<String, Object>> instrumentRates = instrumentRateInsert.execute(createDefaultConnection(), getProjectName()).getRows();
+    }
+
+    @Test
+    public void testSchedule()
+    {
+        goToProjectHome();
+        clickAndWait(Locator.linkWithText("Your project list"));
+        assertTextPresent(PROJECT_1, PROJECT_2);
+        clickAndWait(Locator.linkWithText(PROJECT_1));
+        waitAndClickAndWait(Locator.linkWithText("Schedule instrument time"));
+
+        String yearMonth = Calendar.getInstance().get(Calendar.YEAR) + "-";
+        int month = (Calendar.getInstance().get(Calendar.MONTH) + 1);
+        if (month < 10)
+        {
+            yearMonth += yearMonth;
+        }
+        yearMonth += month;
+
+        scheduleInstrument(yearMonth + "-02");
+        scheduleInstrument(yearMonth + "-03");
+        scheduleInstrument(yearMonth + "-03", true);
+        scheduleInstrument(yearMonth + "-03");
+
+        assertProjectEventCounts(2, 0);
+
+        sleep(1000);  // Wait for the dialog to clear out of the way
+        doAndWaitForPageToLoad(() -> selectOptionByText(Locator.id("projectDropDown"), PROJECT_2));
+
+        scheduleInstrument(yearMonth + "-04");
+        assertProjectEventCounts(1, 2);
+
+        scheduleInstrument(yearMonth + "-05");
+        assertProjectEventCounts(2, 2);
+
+        sleep(1000);  // Wait for the dialog to clear out of the way
+        doAndWaitForPageToLoad(() -> selectOptionByText(Locator.id("instrumentDropDown"), INSTRUMENT_2));
+        scheduleInstrument(yearMonth + "-06");
+        assertProjectEventCounts(1, 0);
+
+        goToDashboard();
+        clickAndWait(Locator.linkWithText("All instrument calendar view"));
+        assertTextPresent(INSTRUMENT_1, INSTRUMENT_2, INACTIVE_INSTRUMENT);
+
+        waitForElementToBeVisible(Locator.tagWithClass("div", "activeProjectEvent"));
+        assertProjectEventCounts(5, 0);
+
+        selectOptionByText(Locator.id("projectFilter"), PROJECT_2);
+        assertProjectEventCounts(3, 2);
+
+        goToDashboard();
+        clickAndWait(Locator.linkWithText("Instrument billing report"));
+        assertTextPresent("$2,450.00", 4);
+        assertTextPresent("$2,690.00", 1);
+
+        // Future test cases:
+        // Split payment across multiple methods
+        // Schedule for hours within a day instead of 24-hour periods
+        // Check billing for individual months, including reservations that span month boundaries with start/end dates
+        // Ensure that overlapping reservations are rejected
+        // Ensure that reservations cannot be made for inactive instruments
+    }
+
+    private void assertProjectEventCounts(int expectedActiveCount, int expectedOtherCount)
+    {
+        assertElementPresent(Locator.tagWithClass("div", "activeProjectEvent"), expectedActiveCount);
+        assertElementPresent(Locator.tagWithClass("div", "otherProjectEvent"), expectedOtherCount);
```

Comment:
Alan recently made the argument that we should be using `Locator.byClass("activeProjectEvent")` for robustness. I don't think this project needs to worry about that, but wanted to raise it.

Reply:
Sure, I can avoid using the tag though I am not anticipating changing this to be a `<h2>` or something